### DISCUSSION TITLE
Bond Zero coupon curve from Bond Par curve?

### DISCUSSION THREAD

**Discussion by mariobajo** [domokane/FinancePy#198] [Q&A]
Hi, is there a way to derive zero coupon rates from par rates? I have tried but I am not sure the the output I get is right... I am doing this for Tsy bond par rates, not for depo & swap rates. Thanks!!!

**Comment by heykav** [comment]
Yes — `BondBootstrapDiscountCurve` in `financepy.market.curves` does exactly this: given a list of coupon-bearing `Bond` objects and their market clean prices, it root-finds a discount factor per maturity so each bond exactly reprices to its quoted clean price, then you read zero rates off the resulting curve. That's a genuine bootstrap (exact refit at each maturity), not a parametric fit.

For Treasuries specifically, use `BondMarkets.UNITED_STATES` for the day-count/frequency conventions:

```python
from financepy.products.bonds.bond import Bond
from financepy.products.bonds import get_bond_market_conventions, BondMarkets
from financepy.market.curves import BondBootstrapDiscountCurve
from financepy.utils.date import Date

acc_dc_type, freq_type, spot_days, ex_div_days, cal = get_bond_market_conventions(
    BondMarkets.UNITED_STATES
)

today = Date(18, 9, 2024)
settle_dt = today.add_weekdays(spot_days)

bonds = []
clean_prices = []
for issue_dt, maturity_dt, coupon, clean_price in your_par_bond_data:
    bond = Bond(
        issue_dt, maturity_dt, coupon, freq_type, acc_dc_type,
        ex_div_days=ex_div_days, cal_type=cal,
    )
    bonds.append(bond)
    clean_prices.append(clean_price)

curve = BondBootstrapDiscountCurve(settle_dt, bonds, clean_prices)

zero_rates = curve.zero_rate([b.maturity_dt for b in bonds])
```

`examples/scripts/bonds/example_bond_bootstrap_discount_curve.py` in the repo has a fuller worked version (it uses UK gilt conventions there, but swapping in `BondMarkets.UNITED_STATES` is the only change needed for Treasuries) — worth a look if you want to see the interpolation-type comparison it also does, since `BondBootstrapDiscountCurve` takes an `interp_type` argument that affects the shape of the curve between your bootstrapped nodes, not just at them.

One likely source of the "not sure the output is right" feeling: if your par curve has widely spaced maturities, the choice of `interp_type` (flat forward vs. others) can make zero rates between bootstrap nodes look surprising even though the bonds at the nodes themselves reprice exactly — worth checking `curve.check_refit_flag=True` (constructor arg) if you haven't, since that explicitly verifies each bond reprices to its input clean price and will tell you if something upstream (day counts, coupon dates) is off rather than the bootstrap itself.
